src/lib.rs

```diff
@@ -710,26 +746,48 @@ impl SystemHandle {
     /// Subscribe given `recipient` to events of type `E`. See [`Context::subscribe()`].
     pub fn subscribe_recipient<M: 'static, E: Event + Into<M>>(&self, recipient: Recipient<M>) {
         let mut event_subscribers = self.event_subscribers.write();
+        event_subscribers.subscribe_recipient::<M, E>(recipient);
+    }
 
-        let subs = event_subscribers.events.entry(TypeId::of::<E>()).or_default();
+    /// Subscribe given `recipient` to events of type `E` and send the last cached event to it.
+    /// See [`Context::subscribe_and_receive_latest()`].
+    pub fn subscribe_and_receive_latest<M: 'static, E: Event + Into<M>>(
+        &self,
+        recipient: Recipient<M>,
+    ) -> Result<(), SendError> {
+        let mut event_subscribers = self.event_subscribers.write();
 
-        subs.push(Box::new(move |e| {
-            if let Some(event) = e.downcast_ref::<E>() {
-                let msg = event.clone();
-                recipient.send(msg.into())?;
+        // Send the last cached value if there is one. The cached event sending and adding ourselves
+        // to the subscriber list needs to be under the same write lock guard to avoid race
+        // conditions between this method and `publish()` (to guarantee exactly-once delivery).
```

Comment:
It seems to me that `self.event_subscribers` usage is very correlated with `last_value_cache` usage. Wouldn't it be better to put both the subscribers hashmap and the cache dashmap behind a single RwLock?

Reply:
Well, they are already under a single `RwLock`, in
https://github.com/tonarino/actor/blob/ba3dba1225d04105bf8a8037ec9ee1d11f47f595/src/lib.rs#L263
https://github.com/tonarino/actor/blob/ba3dba1225d04105bf8a8037ec9ee1d11f47f595/src/lib.rs#L232-L239
..? (with slightly misleading variable/struct name)

The motivation for the comments is more that it would be possible to put `last_value_cache` outside of the lock, but we don't want it.

Reply:
OIC, I completely misunderstood the comments, sorry!